### DISCUSSION TITLE
Is it possible to use persistent connections (HTTP keep alive) in React Native?

### DISCUSSION THREAD

**Discussion by mikescott** [axios/axios#6310] [Q&A]
Hi,

My backend has HTTP2, HTTP3, and therefore keep alive, enabled. Browsers making API requests complete the request 100 to 150ms faster than my React Native app, and I believe that's because the app establishes a new TLS connection for every request, unlike in the browser where it seems to be reuisng an existing connection.

Is there any way to make Axios re-use connections in React Native? 

I've seen an example where that can be done inside node.js, but not RN.

**Comment by jasonsaayman** [comment]
Axios cannot directly enable or configure HTTP keep-alive in React Native the way it can in Node.js.

The Node.js examples usually use `httpAgent` / `httpsAgent`:

```js
import http from 'node:http';
import https from 'node:https';

axios.create({
  httpAgent: new http.Agent({ keepAlive: true }),
  httpsAgent: new https.Agent({ keepAlive: true })
});
```

That only works with Axios’ Node http adapter. In React Native, Axios uses the XHR-style adapter and delegates networking to React Native’s native networking stack. So httpAgent, httpsAgent, and Node socket pooling options do not apply.

In practice, connection reuse in React Native is controlled by the native layer:
- iOS: usually NSURLSession
- Android: usually OkHttp

Those native clients may already reuse connections, but Axios does not expose a JS-level option to force or tune that behaviour.
A few things to check:
- Make sure every request uses the exact same scheme, host, and port.
- Check whether the server or proxy sends Connection: close.
- Confirm HTTP/2 is actually negotiated from the React Native app, not only from browsers.
- Compare with React Native fetch() from the same app. If fetch() has the same latency, the issue is below Axios.
- Use server logs, a proxy, or packet capture to verify whether a new TLS handshake is happening on every request.
- Do not try to set Connection: keep-alive manually. It is not useful for HTTP/2 and may be ignored or blocked by the platform.

If you need explicit control over connection pooling in React Native, that usually requires native configuration rather than Axios configuration. For example, on Android, you may need to customise the app’s OkHttpClient; on iOS, you may need to customise the networking layer around NSURLSession.

So the short answer is: Axios itself cannot make React Native reuse connections. React Native’s native networking stack is responsible for that.
